src/signalrclient/signalr_default_scheduler.cpp

```diff
@@ -0,0 +1,217 @@
+                                        return closed || callback != nullptr;
+                                    });
+                            }
+
+                            if (closed && callback == nullptr)
+                            {
+                                assert(callback == nullptr);
+                                return;
+                            }
+
+                            cb = callback;
+                            internals->m_callback = nullptr;
+                        } // unlock
+
+                        try
+                        {
+                            cb();
+                        }
+                        catch (...)
+                        {
+                            // ignore exceptions?
+                        }
+                    } // destruct cb, otherwise it's possible a shared_ptr is being held by the lambda/function and on destruction it could schedule work which could be
+                    // added to this thread and if it then blocks waiting for completion it would deadlock. If we destruct before setting the m_busy flag,
+                    // another thread will run the work instead.
+
+                    internals->m_busy = false;
+                }
+
+                assert(internals->m_callback == nullptr);
+            });
+    }
+
+    void thread::add(signalr_base_cb cb)
+    {
+        {
+            assert(m_internals->m_closed == false);
+            assert(m_internals->m_busy == false);
+
+            std::lock_guard<std::mutex> lock(m_internals->m_callback_lock);
+            m_internals->m_callback = cb;
+            m_internals->m_busy = true;
+        } // unlock
+        m_internals->m_callback_cv.notify_one();
+    }
+
+    void thread::shutdown()
+    {
+        {
+            std::unique_lock<std::mutex> lock(m_internals->m_callback_lock);
+            m_internals->m_closed = true;
+        }
+        m_internals->m_callback_cv.notify_one();
+        m_thread.join();
+    }
+
+    bool thread::is_free() const
+    {
+        return !m_internals->m_busy;
+    }
+
+    thread::~thread()
+    {
+        shutdown();
+    }
+
+    void signalr_default_scheduler::schedule(const signalr_base_cb& cb, std::chrono::milliseconds delay)
+    {
+        {
+            assert(m_internals->m_closed == false);
+            std::lock_guard<std::mutex> lock(m_internals->m_callback_lock);
+            m_internals->m_callbacks.push_back(std::make_pair(cb, std::chrono::steady_clock::now() + delay));
+        } // unlock
+
+        // If callback has a delay, no point in running the logic to check if it should run early
+        if (delay == std::chrono::milliseconds::zero())
+        {
+            m_internals->m_callback_cv.notify_one();
+        }
+    }
+
+    void signalr_default_scheduler::run()
+    {
+        auto internals = m_internals;
+
+        std::thread([=]()
+            {
+                std::vector<thread> threads{ 5 };
```

Comment:
Same here for thread count, probably should be configurable.

Reply:
Let's add this stuff to https://github.com/dotnet/aspnetcore/issues/8716

Reply:
Added https://github.com/dotnet/aspnetcore/issues/8716#issuecomment-784357610